### ISSUE TITLE
[Bug] [Qwen3-Next] RMSNorm2ZeroCenteredRMSNormMapping: In-place mutation corrupts source model

### ISSUE THREAD

**Issue by w2ais** [NVIDIA-NeMo/Megatron-Bridge#1929]
---
name: Bug report
about: In-place tensor mutation corrupts source model weights
title: "RMSNorm2ZeroCenteredRMSNormMapping: In-place mutation corrupts source model"
labels: bug
assignees: ''

---

**Describe the bug**

`RMSNorm2ZeroCenteredRMSNormMapping` uses in-place tensor operations (`.data = ...`) during weight conversion, which corrupts the source model. The corruption accumulates with each conversion call. This is particularly problematic with RL Weight Sync, leading to immediate model corruption. This bug impacts Qwen3-Next.

**Affected file:** `src/megatron/bridge/models/conversion/param_mapping.py:2061-2077`

**Steps/Code to reproduce bug**

```python
import torch
from megatron.bridge.models.conversion.param_mapping import (
    RMSNorm2ZeroCenteredRMSNormMapping,
    AutoMapping
)

# Simulate Megatron weight tensor
megatron_weight = torch.zeros(128)  # mean = 0.0
print(f"Before export: mean = {megatron_weight.mean():.1f}")

# Create mapping and mock module
mapping = RMSNorm2ZeroCenteredRMSNormMapping(
    megatron_param="decoder.layers.0.self_attention.out_norm.weight",
    hf_param="model.layers.0.linear_attn.norm.weight",
)

# Mock module - register it as replicated type so AutoMapping detection works
class MockModule:
    tensor_model_parallel = False  # This makes it "replicated"

# Register the module type to avoid detection error
AutoMapping.register_module_type('MockModule', 'replicated')

# First export
_ = mapping.megatron_to_hf(megatron_weight, MockModule())
print(f"After export 1: mean = {megatron_weight.mean():.1f}")  # Now 1.0!

# Second export
_ = mapping.megatron_to_hf(megatron_weight, MockModule())
print(f"After export 2: mean = {megatron_weight.mean():.1f}")  # Now 2.0!
```

Output:
```
Before export: mean = 0.0
After export 1: mean = 1.0
After export 2: mean = 2.0
```

**Expected behavior**

Export and load operations should be **read-only** on the source model. The source weights should remain unchanged after conversion.

```
Before export: mean = 0.0
After export 1: mean = 0.0  # Unchanged
After export 2: mean = 0.0  # Unchanged
```

**Additional context**

**Root cause**

Both methods modify tensors in-place using `.data` assignment:

```python
# hf_to_megatron (line 2066-2068)
def hf_to_megatron(self, hf_weights, megatron_module):
    hf_weights.data = hf_weights.data - 1  # Mutates input tensor!
    return super().hf_to_megatron(hf_weights, megatron_module)

# megatron_to_hf (line 2070-2077)
def megatron_to_hf(self, megatron_weights, megatron_module):
    hf_weights = super().megatron_to_hf(megatron_weights, megatron_module)
    inner_hf_weight = list(hf_weights.values())[0]
    inner_hf_weight.data = inner_hf_weight.data + 1  # Mutates source tensor!
    return hf_weights
```

The `.data` attribute assignment bypasses PyTorch's copy-on-write semantics and directly modifies the underlying storage, affecting all references to the tensor.

**Suggested fix**

Replace in-place operations with tensor creation:

```python
def hf_to_megatron(self, hf_weights, megatron_module):
    hf_weights = hf_weights.clone()
    hf_weights.data -= 1
    return super().hf_to_megatron(hf_weights, megatron_module)

def megatron_to_hf(self, megatron_weights, megatron_module):
    hf_weights = super().megatron_to_hf(megatron_weights, megatron_module)
    key = list(hf_weights.keys())[0]
    value = hf_weights[key].clone()
    value.data += 1
    return {key: value}
```

**Impact**

- Export during training corrupts live model weights
- Multiple exports cause cumulative corruption
- Round-trip conversions corrupt weights
- Affects all models using `RMSNorm2ZeroCenteredRMSNormMapping` (currenlyt seems to be only Qwen3-Next)

**Comment by w2ais** [comment]
PR #1930

**Comment by yaoyu-33** [comment]
thanks for contribution @w2ais . approved your pr.

**Comment by JRD971000** [comment]
Fixed by [#1930](https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/1930) (merged 2026-01-14). Closing

**Comment by pruprakash** [comment]
## QA RCCA Analysis

### 1. Fix Reference
- **PR**: #1930 - [Qwen3-Next] fix in-place mutation weight corruption (merged 2026-01-14, community contribution)
- **Issue**: #1929 - RMSNorm2ZeroCenteredRMSNormMapping: In-place mutation corrupts source model

### 2. Root Cause
`RMSNorm2ZeroCenteredRMSNormMapping` applied the zero-centered offset with in-place ops on the SOURCE tensors during conversion (`-1` on hf→megatron, `+1` on megatron→hf). Every conversion shifted the live weights, and repeated conversion (RL weight sync per step) corrupted the model cumulatively. PR #1930 clones before applying the offset in both directions (current `param_mapping.py` carries `.clone()` on both paths).

### 3. Trigger Configuration
Any Qwen3-Next conversion/export cycle repeated on the same live model — most acutely RL weight-sync loops.

### 4. Nature of the Bug
**Classification**: CODE BUG - silent weight corruption (in-place mutation)

### 5. Existing Test Coverage
**In Fix PR**: NO - no test accompanied the contribution.
**In NMFW Tests**: GAP before this analysis - upstream bridge tests only assert the mapping EXISTS in the Qwen3-Next registry, not the non-mutation property.

### 6. Coverage Assessment
| Test | What it covers | Covers this bug? |
|-----------|--------|------------|
| `tests/unit_tests/models/qwen/test_qwen3_next_bridge.py` (upstream) | mapping present in registry | No (existence only) |
| `testcases/rcca/models/test_rmsnorm_zerocentered_no_mutation.py` (NMFW, NEW) | source tensors unchanged across repeated conversions in both directions; outputs correct (±1) and stable | Yes |

### 7. New Regression Test
**ADDED** to NMFW test suite:
- `testcases/rcca/models/test_rmsnorm_zerocentered_no_mutation.py`
- Tests: `test_hf_to_megatron_does_not_mutate_source`, `test_megatron_to_hf_does_not_mutate_source` - exercise the exact fixed methods (parent transport stubbed to isolate the offset logic, including the parent-aliases-source worst case) and assert the source stays bit-identical across repeated calls; verified against container nvcr.io/nvidian/nemo:26.06.01.rc0 with latest main.

### 8. Conclusion
**Verdict**: GAP COVERED - TEST ADDED (NMFW test suite)
